File: ruby/approx_pi_finder.rb
require File.dirname(__FILE__) + '/point'

class ApproxPiFinder
  RADIUS = 1000.0

  attr_reader :points, :sample_size

  def initialize(sample_size)
    @sample_size = sample_size
  end

  def distances
    @distances ||= points.map { |point| point.distance }
  end

  def points
    @points ||= 1.upto(sample_size).map { Point.new(rand * RADIUS, rand * RADIUS) }
  end

  def inside_circle
    distances.count {|d| d < RADIUS}
  end

  def approx_pi
    @approx_pi ||= (4.0 * inside_circle) / sample_size
  end

  def error
    (approx_pi - Math::PI) * 100.0 / Math::PI
  end

  def output_to_console
    $stdout.puts "Approx pi : #{approx_pi}"
    $stdout.puts "Actual pi : #{Math::PI}"
    $stdout.puts "Error     : #{error}%"
  end
end

File: go/approx_pi_finder.go
package main

import (
	"fmt"
	"math"
	"math/rand"
)

const (
	radius = 1000.0
)

func randomTill(limit float64) float64 {
	return rand.Float64() * limit
}

type ApproxPiFinder struct {
	sampleSize int
	points     []Point
}

func newApproxPiFinder(sampleSize int) ApproxPiFinder {
	return ApproxPiFinder{
		sampleSize: sampleSize,
		points:     make([]Point, sampleSize),
	}
}

func (finder *ApproxPiFinder) assignPoints() {
	for i := 0; i < finder.sampleSize; i++ {
		x := randomTill(radius)
		y := randomTill(radius)
		finder.points[i] = newPoint(x, y)
	}
}

func (finder *ApproxPiFinder) totalPointsInsideCircle() int {
	count := 0
	for _, point := range finder.points {
		if point.distance() < radius {
			count++
		}
	}
	return count
}

func (finder *ApproxPiFinder) approxPi() float64 {
	return 4.0 * float64(finder.totalPointsInsideCircle()) / float64(finder.sampleSize)
}

func (finder *ApproxPiFinder) error() float64 {
	return (finder.approxPi() - math.Pi) * 100.0 / math.Pi
}

func (finder *ApproxPiFinder) evaluate() {
	finder.assignPoints()
}

func (finder *ApproxPiFinder) outputToConsole() {
	fmt.Println()
	fmt.Printf("Approx pi  : %f\n", finder.approxPi())
	fmt.Printf("Actual pi  : %f\n", math.Pi)
	fmt.Printf("Error      : %f%%\n", finder.error())
}
